quixstreams/sources/base.py

```diff
@@ -0,0 +1,375 @@
+import dataclasses
+import threading
+import logging
+
+from abc import ABC, abstractmethod
+from typing import Optional, Union
+
+
+from quixstreams.models.messages import KafkaMessage
+from quixstreams.models.topics import TimestampExtractor, TopicConfig, Topic
+from quixstreams.models.types import Headers
+from quixstreams.rowproducer import RowProducer
+from quixstreams.models.serializers import (
+    BytesSerializer,
+    BytesDeserializer,
+    SerializerType,
+    DeserializerType,
+)
+from quixstreams.checkpointing.exceptions import CheckpointProducerTimeout
+
+
+logger = logging.getLogger(__name__)
+
+__all__ = (
+    "BaseSource",
+    "Source",
+    "CheckpointingSource",
+    "PollingSource",
+    "PollingSourceShutdown",
+)
+
+
+@dataclasses.dataclass()
+class SourceTopic:
+    """
+    Source topic configuration
+
+    See :class:`quixstrems.models.Topic` for more detailts.
+    """
+
+    name: str
+    config: Optional[TopicConfig] = None
+    value_deserializer: Optional[DeserializerType] = "json"
+    key_deserializer: Optional[DeserializerType] = BytesDeserializer()
+    value_serializer: Optional[SerializerType] = "json"
+    key_serializer: Optional[SerializerType] = BytesSerializer()
+    timestamp_extractor: Optional[TimestampExtractor] = None
+
+    def asdict(self):
+        return {
+            field.name: getattr(self, field.name) for field in dataclasses.fields(self)
+        }
+
+
+class BaseSource(ABC):
```

Comment:
Why can't the `BaseSource` and `Source` be merged into a single ABC (called `Source`)?

Are some sources supposed to subclass from lower-level `BaseSource` and some from higher-level `Source`? If yes then what would be the use cases for this distinction?

I think for simplicity we should create and promote a single base class. Of course if possible, I may not be aware of some details.

Reply:
iirc Daniil preference if for one BaseClass that defines the strict minimum implementation. Then we can have a bunch of additional classes that implement additional features.

Reply:
Let's try to find an answer to my question: Are some sources supposed to subclass from lower-level `BaseSource` and some from higher-level `Source`?

Because if the answer is no, and all the final Sources are supposed to derive from Source then I really see no point in having two base classes.

If the answer is yes, then the name `Source` is inappropriate (too generic).

I will tag @daniil-quix here too.

Reply:
> Let's try to find an answer to my question: Are some sources supposed to subclass from lower-level BaseSource and some from higher-level Source?

Sorry, yes `BaseSource` is all you need to implement any source. `Source` is there as an intermediary step to make it easier.

 I'd expect most sources to subclass `Source` and not `BaseSource`, that's why this name is more generic

Reply:
I don't support this approach. My argument against this is that the user will be forced to put extra thought into deciding which class should be the base for his/her new source. That's why I'm asking for examples of sources that would subclass from `Source` and those that would subclass from `BaseSource` directly.

Of course, this is not a blocking disagreement 😉

Reply:
I think it's okay to have the `BaseSource` as a minimal interface definition and the `Source` class, which provides some basic implementation.
We need to mention in the `BaseSource`'s docstring and docs that in most cases users need to subclass `Source` to implement their own.